src/mame/nabu/nabupc.cpp

```diff
@@ -0,0 +1,509 @@
+ *    Q1' -|7      10|-  I0'   | (h)  L  x  x  L  H  H  H  H  H  | L  H  L  H  H  |
+ *         |         |         | (i)  L  x  L  H  H  H  H  H  H  | L  L  H  H  H  |
+ *   GND  -|8       9|-  Q0'   | (j)  L  L  H  H  H  H  H  H  H  | L  H  H  H  H  |
+ *         |         |         +---------------------------------+----------------+
+ *         +---------+
+ * </PRE>
+ */
+static __inline f9318_out_t f9318(f9318_in_t in)
+{
+	int out;
+
+	if (in & PRIO_IN_EI) {
+		out = PRIO_OUT_EO | PRIO_OUT_GS | PRIO_OUT_QZ;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (0 == (in & PRIO_I7)) {
+		out = PRIO_OUT_EO;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I7 == (in & PRIO_I6_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q0;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I6_I7 == (in & PRIO_I5_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q1;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I5_I7 == (in & PRIO_I4_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q0 | PRIO_OUT_Q1;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I4_I7 == (in & PRIO_I3_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q2;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I3_I7 == (in & PRIO_I2_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q0 | PRIO_OUT_Q2;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I2_I7 == (in & PRIO_I1_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q1 | PRIO_OUT_Q2;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	if (PRIO_I1_I7 == (in & PRIO_I0_I7)) {
+		out = PRIO_OUT_EO | PRIO_OUT_Q0 | PRIO_OUT_Q1 | PRIO_OUT_Q2;
+		return static_cast<f9318_out_t>(out);
+	}
+
+	out = PRIO_OUT_QZ | PRIO_OUT_GS;
+	return static_cast<f9318_out_t>(out);
+}
+
+//**************************************************************************
+//  INPUT PORTS
+//**************************************************************************
+
+// CONFIG Input Port
+static INPUT_PORTS_START( nabupc )
+	PORT_START("CONFIG")
+	PORT_CONFNAME( 0x03, 0x02, "BIOS Size" )
+	PORT_CONFSETTING( 0x02, "4k BIOS" )
+	PORT_CONFSETTING( 0x01, "8k BIOS" )
+INPUT_PORTS_END
+
+
+//**************************************************************************
+//  Peripheral Devices
+//**************************************************************************
+
+// HCCA Devices
+static void hcca_devices(device_slot_interface &device)
+{
+	device.option_add("pty",           PSEUDO_TERMINAL);
+	device.option_add("null_modem",    NULL_MODEM);
+	device.option_add("hcca_adapter",  NABUPC_NETWORK_ADAPTER);
+}
+
+// Keyboard Devices
+static void keyboard_devices(device_slot_interface &device)
+{
+	device.option_add("nabu", NABUPC_KEYBOARD);
+	device.option_add("nabu_hle", NABUPC_HLE_KEYBOARD);
+}
+
+// Printer Devices
+static void printer_devices(device_slot_interface &device)
+{
+	device.option_add("ex800", EPSON_EX800);
+	device.option_add("lx800", EPSON_LX800);
+	device.option_add("lx810l", EPSON_LX810L);
+	device.option_add("printer", CENTRONICS_PRINTER);
+}
```

Comment:
Is there a good reason not to use default Centroinix port devices?  There’s plenty of places we allow devices to be plugged in when there’s no driver for them.